code/game/machinery/doppler_array.dm

```diff
@@ -11,24 +11,46 @@ GLOBAL_LIST_EMPTY(doppler_arrays)
 	var/integrated = FALSE
 	var/max_dist = 150
 	verb_say = "states coldly"
+	var/list/message_log = list()
+	var/output_log
 
 /obj/machinery/doppler_array/Initialize()
 	. = ..()
 	GLOB.doppler_arrays += src
 
 /obj/machinery/doppler_array/ComponentInitialize()
+	. = ..()
 	AddComponent(/datum/component/simple_rotation,ROTATION_ALTCLICK | ROTATION_CLOCKWISE,null,null,CALLBACK(src,.proc/rot_message))
 
 /obj/machinery/doppler_array/Destroy()
 	GLOB.doppler_arrays -= src
 	return ..()
 
-/obj/machinery/doppler_array/examine(mob/user)
-	..()
-	to_chat(user, "<span class='notice'>Its dish is facing to the [dir2text(dir)].</span>")
 
-/obj/machinery/doppler_array/process()
-	return PROCESS_KILL
+/obj/machinery/doppler_array/ui_interact(mob/user)
+	. = ..()
+	var/dat = {"<html><head><title>[src.name]</title><style>h3 {margin: 0px; padding: 0px;}</style></head><body><br>

```

Comment:
Please. Use the browser datum.

Reply:
You mean slow TGUI?

Reply:
No i mean browser datum.

Reply:
I'll try that

Reply:
While you're at it make sure to use list.Join() instead of adding each string separately.

Reply:
The html/head/title/body and css are unnecessary now they're created by browser datum already

Reply:
Will.address